### PR TITLE
expression: add builtin function json_valid

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve? 
Add ```json_valid``` builtin function described [#7546](https://github.com/pingcap/tidb/issues/7546).

### What is changed and how it works?
See [here](https://dev.mysql.com/doc/refman/5.7/en/json-attribute-functions.html#function_json-valid) for details.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_json.go [hunk 1]
```diff
@@ -716,7 +720,66 @@ type jsonValidFunctionClass struct {
...
+	baseBuiltinFunc
+}
+
+func (b *builtinJSONValidJsonSig) Clone() builtinFunc {
+	newSig := &builtinJSONValidJsonSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinJSONValidJsonSig) evalInt(row chunk.Row) (res int64, isNull bool, err error) {
```
Comment (SunRunAway) [inline expression/builtin_json.go]: We can add the mysql documentation link as a comment here.

File: expression/builtin_json.go [hunk 2]
```diff
@@ -716,7 +720,66 @@ type jsonValidFunctionClass struct {
...
+	baseBuiltinFunc
+}
+
+func (b *builtinJSONValidStringSig) Clone() builtinFunc {
+	newSig := &builtinJSONValidStringSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinJSONValidStringSig) evalInt(row chunk.Row) (res int64, isNull bool, err error) {
```
Comment (SunRunAway) [inline expression/builtin_json.go]: We can add the mysql documentation link as a comment here.

File: expression/builtin_json.go [hunk 3]
```diff
@@ -716,7 +720,70 @@ type jsonValidFunctionClass struct {
...
+
+	var sig builtinFunc
+	argType := args[0].GetType().EvalType()
+	switch argType {
+	case types.ETJson:
+		bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, types.ETJson)
+		sig = &builtinJSONValidJSONSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_JsonValidJsonSig)
+	default:
+		bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, argType)
```
Comment (zz-jason) [inline expression/builtin_json.go]:
```suggestion
		bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, types.ETString)
```

Comment (zz-jason) [reply expression/builtin_json.go]: If its argument is not a string or a json, should we convert it to a string?

Comment (Reminiscent) [reply expression/builtin_json.go]: In mysql5.7
```
mysql> select json_valid(0);
+---------------+
| json_valid(0) |
+---------------+
|             0 |
+---------------+
1 row in set (0.00 sec)

mysql> select json_valid("0");
+-----------------+
| json_valid("0") |
+-----------------+
|               1 |
+-----------------+
1 row in set (0.00 sec)
```

**Comment by Reminiscent** [comment]
@SunRunAway  @zz-jason @qw4990  PTAL

File: expression/builtin_json.go (hunk 3 above)
Comment (zz-jason) [reply expression/builtin_json.go]: Should we handle int/double/decimal values separately?

Comment (Reminiscent) [reply expression/builtin_json.go]: See [here](https://github.com/mysql/mysql-server/blob/8.0/sql/item_json_func.cc#L295-L319) for more details.

**Comment by Reminiscent** [comment]
cherry-pick to release-3.0 in PR #13133 
cherry-pick to release-3.1 in PR #13138

**Pull request merged**